File: backend/app/schemas/dag_summary.py
"""Pydantic DTOs for DAG summary/statistics endpoint."""

from pydantic import BaseModel


class DagTaskSummary(BaseModel):
    task_id: str
    pipeline_name: str | None = None
    pipeline_id: str | None = None
    status: str
    latest_duration_seconds: float | None = None
    avg_duration_seconds: float | None = None
    task_group_id: str | None = None


class DagSummary(BaseModel):
    dag_id: str
    description: str | None = None
    schedule_interval: str | None = None
    is_paused: bool = False
    task_count: int = 0
    pipeline_count: int = 0

    # Duration aggregates (from latest DAG run)
    total_duration_seconds: float | None = None
    avg_task_duration_seconds: float | None = None
    min_task_duration_seconds: float | None = None
    max_task_duration_seconds: float | None = None

    # Status counts (from current Airflow statuses) — dynamic keys, one per Airflow state
    status_counts: dict[str, int] = {}
    success_rate: float | None = None

    # Timing
    latest_run_start: str | None = None
    latest_run_end: str | None = None
    typical_finish_hour: str | None = None

    # Period-based history (default 30d, adjusts with date range)
    total_runs_30d: int = 0
    dag_success_rate_30d: float | None = None
    period_label: str = "30d"

    tasks: list[DagTaskSummary] = []


class DagSummaryAggregate(BaseModel):
    total_dags: int = 0
    total_pipelines: int = 0
    active_dags: int = 0
    overall_success_rate: float | None = None
    total_runs_30d: int = 0
    period_label: str = "30d"


class DagSummaryResponse(BaseModel):
    aggregate: DagSummaryAggregate
    dags: list[DagSummary] = []

File: frontend/src/types/dag-summary.ts
export interface DagTaskSummary {
  task_id: string;
  pipeline_name: string | null;
  pipeline_id: string | null;
  status: string;
  latest_duration_seconds: number | null;
  avg_duration_seconds: number | null;
  task_group_id: string | null;
}

export interface DagSummary {
  dag_id: string;
  description: string | null;
  schedule_interval: string | null;
  is_paused: boolean;
  task_count: number;
  pipeline_count: number;
  total_duration_seconds: number | null;
  avg_task_duration_seconds: number | null;
  min_task_duration_seconds: number | null;
  max_task_duration_seconds: number | null;
  status_counts: Record<string, number>;
  success_rate: number | null;
  latest_run_start: string | null;
  latest_run_end: string | null;
  typical_finish_hour: string | null;
  total_runs_30d: number;
  dag_success_rate_30d: number | null;
  tasks: DagTaskSummary[];
}

export interface DagSummaryAggregate {
  total_dags: number;
  total_pipelines: number;
  active_dags: number;
  overall_success_rate: number | null;
  total_runs_30d: number;
}

export interface DagSummaryResponse {
  aggregate: DagSummaryAggregate;
  dags: DagSummary[];
}
